colexec: skip over unneeded columns in cfetcher

asubiotto opened the pull request:
The cfetcher was previously eagerly returning all column types even if they
were not needed. This could result in cases where we would refuse to read from
a table that contained any types not supported by the vectorized engine, even
if we only wanted columns whose types were supported.

Release note: None

I'm not very familiar with this code so let me know if you see a better way of doing this.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42616)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colencoding/key_encoding.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-Lu9s8QK6tQ66jJ5AvvN:-Lu9s8QK6tQ66jJ5AvvO:bw459hs) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colencoding/key_encoding.go#L250)):*
> ```Go
> // skipTableKey skips a value of type valType in key, returning the remainder
> // of the key.
> // TODO(jordan): should use this approach in the normal row fetcher. Wait, why?
> ```

I think it means that we could also skip over parts of the keys when using row fetcher and when not all columns are needed. It would probably require modification of `TableReader` since it seems to me that right now all columns are output, and post processing step renders out unneeded columns.
___
*[pkg/sql/colexec/cfetcher.go, line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-Lu9uVM6--pHRWD6MXW0:-Lu9uVM6--pHRWD6MXW1:b-a7yrgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colexec/cfetcher.go#L227)):*
> ```Go
> 
> 		// outputIdxMap is used to get the output index into colvecs for a given column.
> 		outputIdxMap colIdxMap
> ```

Do we need another map here? I might be wrong, but it seems to me that we can use `colIdxMap` for this purpose. We would need to update `cTableInfo.cols` accordingly to have only the needed columns. Then we would not need to `delete` from the map below.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
cc @rohany you might have some suggestions/comments as well.

rohany reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/colencoding/key_encoding.go, line 250 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-Lu9s8QK6tQ66jJ5AvvN:-Lu9zLQw3c-aAftVpxu-:b-r1jjtp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colencoding/key_encoding.go#L250)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think it means that we could also skip over parts of the keys when using row fetcher and when not all columns are needed. It would probably require modification of `TableReader` since it seems to me that right now all columns are output, and post processing step renders out unneeded columns.
</blockquote></details>

Yes, we should use this skipping approach in the normal fetcher. (i misread this when looking at it the first time).
___
*[pkg/sql/colexec/cfetcher.go, line 227 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-Lu9uVM6--pHRWD6MXW0:-LuA0MuX1VH6RMdQENaN:b-qkhwl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colexec/cfetcher.go#L227)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do we need another map here? I might be wrong, but it seems to me that we can use `colIdxMap` for this purpose. We would need to update `cTableInfo.cols` accordingly to have only the needed columns. Then we would not need to `delete` from the map below.
</blockquote></details>

Agreed -- the `neededValsByColIdx` field contains this information already -- can you use this information when decoding?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42616)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/colencoding/key_encoding.go, line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-LuACF7m5NPoTMW_ahPD:-LuACF7m5NPoTMW_ahPE:bkmn8f1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colencoding/key_encoding.go#L261)):*
> ```Go
> 		return key, nil
> 	}
> 	n, err := encoding.PeekLength(key)
> ```

This doesn't make any sense and will (I hope???) cause many tests to fail.

This part is skipping *key* values, not *value* values - and key values do not have length prefixes! Length prefixes are only possible in the value, where things don't have to be lexicographically ordered.
___
*[pkg/sql/colexec/cfetcher.go, line 1088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-LuACls22YwLzoirKSQp:-LuACls22YwLzoirKSQq:bp5rjki) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colexec/cfetcher.go#L1088)):*
> ```Go
> 		}
> 
> 		// So we need to translate
> ```

nit: this sentence appears to


<!-- Sent from Reviewable.io -->

Commit:
colexec: skip over unneeded columns with unknown types in cfetcher

Previously, we would return an error if we encountered an unknown type in the
cfetcher, even if the values from the column were never decoded into the output
coldata.Batch. This change supports reading from a table with unknown types as
long as those columns are marked as not needed. The approach this patch takes
is to allow a coldata.Batch to be created with unknown types, but disallowing
any uses of unknown type Vecs, other than observing that it is of unknown type.

Release note (sql change): Vectorized queries that execute only on supported
types even if those types form part of a table with unsupported types are now
run through the vectorized engine. This would previously fall back to the
row-by-row execution engine.

asubiotto commented:
Updated after offline discussions to a more contained solution which does not mess with the output shape, which would require a larger refactor to get right. PTAL.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-:-LuJYZVF0FD9H5MzTJKN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/col/coldata/unknown_vec.go, line 11 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-LuJYcVA4tIzvskp2sho:-LuJYcVA4tIzvskp2shp:b-bjtbgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c93810c41182e0a29353ba699131c23d376c596/pkg/col/coldata/unknown_vec.go#L11)):*
> ```Go
> // licenses/APL.txt.
> 
> package coldata
> ```

[nit]: `s/2018/2019/g`.
___
*[pkg/sql/colexec/cfetcher.go, line 1088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-LuACls22YwLzoirKSQp:-LuJYGhS3lZvGaOXofoA:b2ni78k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5446b969aa7611113c8d7d5f9e0dc861f42bfc9/pkg/sql/colexec/cfetcher.go#L1088)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: this sentence appears to
</blockquote></details>

lol
___
*[pkg/sql/colexec/mem_estimation.go, line 70 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42616#-LuJYnBIC2MXVaL6kObx:-LuJYnBIC2MXVaL6kOby:b9zxeo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c93810c41182e0a29353ba699131c23d376c596/pkg/sql/colexec/mem_estimation.go#L70)):*
> ```Go
> 			acc += sizeOfTime
> 		case coltypes.Unhandled:
> 			// Placeholder Vecs of unknown types are allowed.
> ```

[super nit]: `s/Vecs/coldata.Vecs/g`.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Blank comments were:
- s/2018/2019/g
- s/Vecs/coldata.Vecs/g

rohany commented:
LGTM aside from those nits.

asubiotto commented:
bors r=yuzefovich,rohany
